File: smoke_tests/rust/ssrf.rs
// SSRF vulnerability in Rust
use reqwest;

async fn fetch_url_unsafe(url: &str) -> Result<String, reqwest::Error> {
    // VULNERABLE: Fetching user-provided URL without validation
    let response = reqwest::get(url).await?;
    response.text().await
}

async fn proxy_request_unsafe(target_url: &str) -> Result<Vec<u8>, reqwest::Error> {
    // VULNERABLE: Acting as open proxy
    let client = reqwest::Client::new();
    let response = client.get(target_url).send().await?;
    let bytes = response.bytes().await?;
    Ok(bytes.to_vec())
}

fn make_internal_request(host: &str, path: &str) -> String {
    // VULNERABLE: User controls internal request destination
    let url = format!("http://{}:8080{}", host, path);
    url
}

File: smoke_tests/typescript/ssrf.ts
// SSRF vulnerabilities in TypeScript
import axios from 'axios';
import fetch from 'node-fetch';

class SsrfVulnerabilities {
    // VULNERABLE: User-controlled URL in fetch
    async fetchUserUrl(url: string): Promise<string> {
        const response = await fetch(url);
        return response.text();
    }

    // VULNERABLE: User-controlled URL in axios
    async axiosGet(endpoint: string): Promise<any> {
        const response = await axios.get(endpoint);
        return response.data;
    }

    // VULNERABLE: URL from query parameter
    async proxyRequest(targetUrl: string): Promise<Response> {
        return fetch(targetUrl);
    }

    // VULNERABLE: Internal service access
    async getInternalData(host: string, port: number): Promise<any> {
        const url = `http://${host}:${port}/internal/data`;
        return axios.get(url);
    }

    // VULNERABLE: File URL scheme
    async readLocalFile(path: string): Promise<string> {
        const response = await fetch(`file://${path}`);
        return response.text();
    }
}
